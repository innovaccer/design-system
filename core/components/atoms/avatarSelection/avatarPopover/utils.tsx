import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>,
  listRef?: any,
  withSearch?: boolean,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      navigateOptions('up', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'ArrowDown':
      event.preventDefault();
      navigateOptions('down', focusedOption, setFocusedOption, listRef, withSearch);
      break;
    case 'Enter':
      handleEnterKey(focusedOption);
      setHighlightLastItem?.(false);
      setHighlightFirstItem?.(false);
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }
};

const handleEnterKey = (focusedOption: Element | undefined) => {
  (focusedOption as HTMLElement)?.click();
};

const navigateOptions = (
  direction: string,
  focusedOption: Element | undefined,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  const listItems = listRef?.current.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let index = Array.from(listItems).findIndex((item) => {
    return item == focusedOption;
  });

  if (index === -1) {
    index = direction === 'up' ? listItems.length - 1 : 0;
  } else if (
    (withSearch && index === 0 && direction === 'up') ||
    (withSearch && index === listItems.length - 1 && direction === 'down')
  ) {
    const searchInput = listRef.current.querySelector('[data-test="DesignSystem-AvatarSelection--Input"]');
    searchInput.focus();
    setFocusedOption && setFocusedOption(searchInput);
  } else {
    index = direction === 'up' ? (index - 1 + listItems.length) % listItems.length : (index + 1) % listItems.length;

    const targetOption = listItems[index];

    (targetOption as HTMLElement).focus();
    setFocusedOption && setFocusedOption(targetOption);
    targetOption.scrollIntoView({ block: 'center' });
  }
};

export const handleInputKeyDown = (
  event: React.KeyboardEvent,
  listRef: any,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  triggerRef?: any
) => {
  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  switch (event.key) {
    case 'ArrowUp':
      event.preventDefault();
      targetOption = listItems[listItems.length - 1];
      break;
    case 'ArrowDown':
      event.preventDefault();
      targetOption = listItems[0];
      break;
    case 'Escape':
      setOpenPopover?.(false);
      triggerRef.current.focus();
      setFocusedOption?.(undefined);
      break;
    default:
      break;
  }

  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView?.({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
