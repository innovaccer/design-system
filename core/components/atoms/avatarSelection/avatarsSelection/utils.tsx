import * as React from 'react';

export const handleKeyDown = (
  event: React.KeyboardEvent,
  setOpenPopover?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightFirstItem?: React.Dispatch<React.SetStateAction<boolean>>,
  setHighlightLastItem?: React.Dispatch<React.SetStateAction<boolean>>
) => {
  switch (event.key) {
    case 'Enter':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowDown':
      setOpenPopover?.(true);
      setHighlightFirstItem?.(true);
      break;
    case 'ArrowUp':
      setOpenPopover?.(true);
      setHighlightLastItem?.(true);
      break;
    default:
      break;
  }
};

export const focusListItem = (
  position: string,
  setFocusedOption?: React.Dispatch<React.SetStateAction<HTMLElement | undefined>>,
  listRef?: any,
  withSearch?: boolean
) => {
  if (withSearch && position === 'down') {
    const searchInput = listRef.current?.querySelectorAll('[data-test="DesignSystem-AvatarSelection--Input"]');
    searchInput && searchInput[0]?.focus();
    setFocusedOption?.(searchInput);
    return;
  }

  const listItems = listRef.current?.querySelectorAll('[data-test="DesignSystem-Listbox-ItemWrapper"]');
  let targetOption;

  if (position === 'down') {
    targetOption = listItems?.[0];
  } else {
    targetOption = listItems?.[listItems.length - 1];
  }
  (targetOption as HTMLElement)?.focus();
  targetOption?.scrollIntoView?.({ block: 'center' });
  setFocusedOption && setFocusedOption(targetOption);
};
