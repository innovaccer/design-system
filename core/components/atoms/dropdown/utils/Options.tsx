export let dropdownOptions: any[] = [];

for (let i = 1; i <= 50; i++) {
  dropdownOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
  });
}

export const selectedStoryOptions: any[] = [];

for (let i = 1; i <= 10; i++) {
  selectedStoryOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
    selected: i === 2,
  });
}

export const multiSelectedStoryOptions: any[] = [];

for (let i = 1; i <= 10; i++) {
  multiSelectedStoryOptions.push({
    label: `Option${i}`,
    value: `Option${i}`,
    selected: i === 2 || i === 7,
  });
}

export const storyOptions = dropdownOptions.slice(0, 10);

export const storySections = { 0: 'subheading', 5: 'subheading' };

export const subInfoItems = [
  {
    label: 'Option1',
    value: 'Option1',
    subInfo: 'subInfo'
  },
  {
    label: 'Option2',
    value: 'Option2',
    subInfo: 'subInfo'
  },
  {
    label: 'Option3',
    value: 'Option3',
    subInfo: 'subInfo',
  },
];

export const iconItems = [
  {
    label: 'Option1',
    value: 'Option1',
    icon: 'events'
  },
  {
    label: 'Option2',
    value: 'Option2',
    icon: 'events',
  },
  {
    label: 'Option3',
    value: 'Option3',
    icon: 'events',
  },
];
