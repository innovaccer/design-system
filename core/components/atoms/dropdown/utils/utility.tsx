const getSearchedOptions = (options: any, searchTerm: string) => {
  const result = options.filter((option: any) => option.label.toLowerCase().includes(searchTerm.toLowerCase()));
  return result;
};

export const getOptions = (offset: number, limit: number, searchTerm: string, options: any) => {
  const searchedOptions = searchTerm ? getSearchedOptions(options, searchTerm) : options;
  return new Promise(resolve => {
    resolve({
      offset,
      slicedOptions: searchedOptions.slice(offset, offset + limit),
      options: searchedOptions,
    });
  });
};

export const getValuesFromSelectedObj = (selectedArray: any[] = []) => {
  const selectedValues: any[] = [];

  selectedArray.forEach((selectedObj: any) => {
    const { value } = selectedObj;
    selectedValues.push(value);
  });

  return selectedValues;
};

export const getIndexesFromSelectedObj = (selectedArray: any[] = []) => {
  return (Array.from(Array(selectedArray.length).keys()));
};

export const getValuesFromInd = (selectedArray: number[], options: any[]) => {
  const result: any[] = [];
  selectedArray.forEach(ind => {
    result.push(options[ind].value);
  });

  return result;
};

export const getLabelsFromInd = (selectedArray: number[], options: any[]) => {
  const result: string[] = [];
  selectedArray.forEach(ind => {
    result.push(options[ind].label);
  });

  return result;
};
