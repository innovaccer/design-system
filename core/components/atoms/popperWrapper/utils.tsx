// Helper function to convert CSS class names to style objects
export const createStyleFromClass = (className: string) => {
  if (!className) return {};

  // Create a temporary element to get computed styles
  const tempElement = document.createElement('div');
  tempElement.className = className;
  document.body.appendChild(tempElement);

  const computedStyle = window.getComputedStyle(tempElement);
  const style: React.CSSProperties = {};

  // Extract relevant animation properties
  const animation = computedStyle.animation;
  const animationDuration = computedStyle.animationDuration;
  const animationTimingFunction = computedStyle.animationTimingFunction;
  const animationFillMode = computedStyle.animationFillMode;
  const opacity = computedStyle.opacity;
  const transform = computedStyle.transform;

  if (animation && animation !== 'none') {
    style.animation = animation;
  }
  if (animationDuration && animationDuration !== '0s') {
    style.animationDuration = animationDuration;
  }
  if (animationTimingFunction && animationTimingFunction !== 'ease') {
    style.animationTimingFunction = animationTimingFunction;
  }
  if (animationFillMode && animationFillMode !== 'none') {
    style.animationFillMode = animationFillMode;
  }
  if (opacity && opacity !== '1') {
    style.opacity = opacity;
  }
  if (transform && transform !== 'none') {
    style.transform = transform;
  }

  document.body.removeChild(tempElement);
  return style;
};
