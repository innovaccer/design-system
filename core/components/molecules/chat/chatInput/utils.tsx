import { MessageType } from './ChatInput';

export const positionMentionPopup = (
  textareaRef: React.RefObject<HTMLDivElement>,
  setMentionPosition: React.Dispatch<React.SetStateAction<{ top: number; left: number }>>
) => {
  if (!textareaRef.current) return;

  const selection = window.getSelection();
  if (!selection || !selection.rangeCount) return;

  const range = selection.getRangeAt(0);
  const rect = range.getBoundingClientRect();
  const editorRect = textareaRef.current.getBoundingClientRect();

  // Calculate position based on the cursor position
  let top = rect.bottom + window.scrollY;
  let left = rect.left + window.scrollX;

  // Adjust position if it would go off-screen
  if (top + 150 > window.innerHeight) {
    top = rect.top + window.scrollY - 150;
  }
  if (left + 200 > window.innerWidth) {
    left = window.innerWidth - 200;
  }

  // Adjust position relative to the editor
  top -= editorRect.top + window.scrollY - 12;
  left -= editorRect.left + window.scrollX - 8;

  setMentionPosition({ top, left });
};

export const extractMessageData = (textareaRef: React.RefObject<HTMLDivElement>): MessageType[] => {
  const elements = textareaRef.current?.childNodes;

  return Array.from(elements || []).map((item: any) => {
    if (item?.dataset?.type === 'mention') {
      const mentionValue = JSON.parse(item.dataset.content);
      return { type: 'mention', content: mentionValue?.data, nodeName: item.nodeName };
    }

    // handles new line by pressing enter
    else if (item.innerText === '\n' && item.nodeName === 'DIV') {
      return { type: 'newline', content: '\n', nodeName: item.nodeName };
    }

    // handles line break by pressing shift + enter
    else if (item.nodeName === 'BR') {
      return { type: 'linebreak', content: '<br>', nodeName: item.nodeName };
    }

    return { type: 'text', content: item.innerText || item.textContent, nodeName: item.nodeName };
  });
};
