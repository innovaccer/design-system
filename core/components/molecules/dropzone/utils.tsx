const isIe = (userAgent: string) => {
  return userAgent.indexOf('MSIE') !== -1 || userAgent.indexOf('Trident/') !== -1;
};

const isEdge = (userAgent: string) => {
  return userAgent.indexOf('Edge/') !== -1;
};

export const accepts = (file: File, acceptedFiles?: string | string[]) => {
  if (file && acceptedFiles) {
    const acceptedFilesArray = Array.isArray(acceptedFiles) ? acceptedFiles : acceptedFiles.split(',');
    const fileName = file.name || '';
    const mimeType = (file.type || '').toLowerCase();
    const baseMimeType = mimeType.replace(/\/.*$/, '');

    if (!mimeType) {
      return true;
    }

    return acceptedFilesArray.some((type) => {
      const validType = type.trim().toLowerCase();
      if (validType.charAt(0) === '.') {
        return fileName.toLowerCase().endsWith(validType);
      }
      if (validType.endsWith('/*')) {
        return baseMimeType === validType.replace(/\/.*$/, '');
      }
      return mimeType === validType;
    });
  }
  return true;
};

export const isPropagationStopped = (event: any) => {
  if (typeof event.isPropagationStopped === 'function') {
    return event.isPropagationStopped();
  }
  if (typeof event.cancelBubble !== 'undefined') {
    return event.cancelBubble;
  }
  return false;
};

export const isEvtWithFiles = (event: any) => {
  if (!event.dataTransfer) {
    return !!event.target && !!event.target.files;
  }

  return Array.prototype.some.call(
    event.dataTransfer.types,
    (type) => type === 'Files' || type === 'application/x-moz-file'
  );
};

export const isKindFile = (item: null | Record<string, unknown>) => {
  return typeof item === 'object' && item !== null && item.kind === 'file';
};

export const onDocumentDragOver = (event: DragEvent) => {
  event.preventDefault();
};

export const isIeOrEdge = (userAgent = window.navigator.userAgent) => {
  return isIe(userAgent) || isEdge(userAgent);
};

type functype = (event: Event, ...args: unknown[]) => void;

export const composeEventHandlers = (...fns: functype[]) => {
  return (event: Event, ...args: unknown[]) =>
    fns.some((fn) => {
      if (!isPropagationStopped(event) && fn) {
        fn(event, ...args);
      }
      return isPropagationStopped(event);
    });
};

export const reducer = (state: any, action: any) => {
  switch (action.type) {
    case 'focus':
      return {
        ...state,
        isFocused: true,
      };
    case 'blur':
      return {
        ...state,
        isFocused: false,
      };
    case 'openDialog':
      return {
        ...state,
        isFileDialogActive: true,
      };
    case 'closeDialog':
      return {
        ...state,
        isFileDialogActive: false,
      };
    case 'setDraggedFiles':
      /* eslint no-case-declarations: 0 */
      const { isDragActive, draggedFiles } = action;
      return {
        ...state,
        draggedFiles,
        isDragActive,
      };
    case 'setFiles':
      return {
        ...state,
        acceptedFiles: action.acceptedFiles,
        fileRejections: action.fileRejections,
      };
    case 'reset':
      return {
        ...state,
        isFileDialogActive: false,
        isDragActive: false,
        draggedFiles: [],
        acceptedFiles: [],
        fileRejections: [],
      };
    default:
      return state;
  }
};
