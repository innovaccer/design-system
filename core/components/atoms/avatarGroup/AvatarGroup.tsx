import * as React from 'react';
import classNames from 'classnames';
import { BaseProps, extractBaseProps } from '@/utils/types';
import Avatar, { Appearance } from '@/components/atoms/avatar';
import Popover, { Position, ActionType } from '@/components/molecules/popover';
import { Text } from '@/index';

export type Size = 'regular' | 'tiny';

interface AvatarData extends Record<string, any> {
  firstName?: string;
  lastName?: string;
  appearance?: Appearance;
}

interface PopperProps {
  popperRenderer?: (names: AvatarData[]) => JSX.Element;
  appendToBody?: boolean;
  dark?: boolean;
  position?: Position;
  on?: ActionType;
  maxHeight?: number;
  popperClassName?: string;
}

export interface AvatarGroupProps extends BaseProps {
  /**
   * List of `Avatars`
   *
   * <pre className="DocPage-codeBlock">
   * AvatarData: {
   *  firstName?: string;
   *  lastName?: string;
   *  appearance?: Appearance;
   * }
   * </pre>
   *
   */
  list: AvatarData[];
  /**
   * Max `Avatars` to show before +x.
   */
  max: number;
  /**
   * Border color of `Avatars`.
   */
  borderColor: string;
  /**
   * **Popover for +x avatar**
   *
   * <pre className="DocPage-codeBlock">
   * PopperProps: {
   *   popperRenderer?: (names: AvatarData[]) => JSX.Element;
   *   appendToBody?: boolean;
   *   dark?: boolean;
   *   position?: Position;
   *   on?: ActionType;
   *   maxHeight?: number;
   *   popperClassName?: string;
   * }
   *
   * | Name | Description | Default |
   * | --- | --- | --- |
   * | popperRenderer | Callback function to create custom popover content  | |
   * | appendToBody | Appends `Popover` wrapper inside body | true |
   * | dark | Changes background of `Popover` | true |
   * | position | Position to place `Popover` | bottom |
   * | on | Event triggering the `Popover` | hover |
   * | maxHeight | Max height of `Popover Text Wrapper` (does not work in case of custom popperRenderer) | 150 |
   * | popperClassName | Custom classname added to `Popover` | |
   * </pre>
   *
   */
  popoverOptions: PopperProps;
  /**
   * Position to place the tooltip on `Avatars` shown before +x
   */
  tooltipPosition: Position;
}

export const AvatarGroup = (props: AvatarGroupProps) => {
  const {
    max,
    borderColor,
    popoverOptions,
    tooltipPosition,
    list,
    className,
  } = props;

  const {
    popperRenderer,
    maxHeight = 150,
    position = 'bottom',
    on = 'hover',
    dark = true,
    appendToBody = true,
    popperClassName = '',
  } = popoverOptions;

  const baseProps = extractBaseProps(props);

  const extraAvatars = list.length > max ? (list.length - max > 9 ? 9 : list.length - max) : 0;

  const style = {
    borderRadius: '50%',
    backgroundColor: `${borderColor}`,
    border: `var(--spacing-xs) solid ${borderColor}`,
    boxShadow: `0 0 0 var(--spacing-xs) ${borderColor}`,
  };

  const AvatarGroupClass = classNames({
    ['AvatarGroup']: true,
  }, className);

  const popperClass = classNames({
    ['AvatarGroup-Popper']: true,
  }, popperClassName);

  const trigger = (
    <div data-test="DesignSystem-AvatarGroup--TriggerAvatar" style={style}>
      <Avatar
        appearance="secondary"
        firstName="+"
        lastName={`${extraAvatars}`}
        withTooltip={false}
      />
    </div>
  );

  const renderPopper = () => {
    const extraAvatarsList = list.slice(max, list.length);

    if (popperRenderer && typeof renderPopper === 'function') {
      return popperRenderer(extraAvatarsList);
    }

    return (
      <div className="py-6 pr-4 pl-6">
        <div className="AvatarGroup-TextWrapper" style={{ maxHeight }}>
          {
            extraAvatarsList.map((item, ind) => {
              const { firstName = '', lastName = '' } = item;
              const name = `${firstName} ${lastName}`;

              return (
                <Text
                  key={ind}
                  appearance={dark ? 'white' : 'default'}
                  className={ind < extraAvatars - 1 ? 'mb-5' : ''}
                  data-test="DesignSystem-AvatarGroup--Text"
                >
                  {name}
                </Text>
              );
            })
          }
        </div>
      </div>
    );
  };

  const renderAvatars = () => {
    const avatars = list.slice(0, max).map((item, index) => {
      const { appearance, firstName, lastName } = item;
      return (
        <div
          data-test="DesignSystem-AvatarGroup--Avatar"
          className="AvatarGroup-item"
          style={style}
          key={index}
        >
          <Avatar
            appearance={appearance}
            firstName={firstName}
            lastName={lastName}
            withTooltip={true}
            tooltipPosition={tooltipPosition}
          />
        </div>
      );
    });
    return avatars;
  };

  return (
    <div
      data-test="DesignSystem-AvatarGroup"
      {...baseProps}
      className={`${AvatarGroupClass} d-inline-flex`}
    >
      {renderAvatars()}
      {list.length - max > 0 && (
        <Popover
          on={on}
          dark={dark}
          trigger={trigger}
          position={position}
          appendToBody={appendToBody}
          className={popperClass}
        >
          {renderPopper()}
        </Popover>
      )
      }
    </div>
  );
};

AvatarGroup.displayName = 'AvatarGroup';
AvatarGroup.defaultProps = {
  max: 2,
  borderColor: 'var(--white)',
  tooltipPosition: 'bottom',
  popoverOptions: {},
};

export default AvatarGroup;
