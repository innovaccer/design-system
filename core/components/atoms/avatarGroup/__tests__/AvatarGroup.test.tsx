import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import AvatarGroup, { AvatarGroupProps as Props } from '../AvatarGroup';
import { testHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/testHelper';

export const list = [
  {
    firstName: 'John',
    lastName: 'Doe',
  },
  {
    firstName: 'Steven',
    lastName: 'Packton',
  },
  {
    firstName: 'Nancy',
    lastName: 'Wheeler',
  },
  {
    firstName: 'Monica',
    lastName: 'Geller',
  },
  {
    firstName: 'Arya',
    lastName: 'Stark',
  },
  {
    firstName: 'Rachel',
    lastName: 'Green',
  },
];

describe('AvatarGroup component', () => {
  const mapper = {
    list: valueHelper(list, { required: true }),
  };

  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as Props;

    it(testMessageHelper(attr), () => {
      const tree = render(<AvatarGroup {...attr} />);
      expect(tree).toMatchSnapshot();
    });
  };

  testHelper(mapper, testFunc);
});

describe('AvatarGroup component', () => {
  it('renders all avatars without +x avatar', () => {
    const max = list.length;

    const { getAllByTestId, queryByTestId } = render(<AvatarGroup list={list} max={max} />);
    expect(getAllByTestId('DesignSystem-AvatarGroup--Avatar')).toHaveLength(list.length);
    expect(queryByTestId('DesignSystem-AvatarGroup--TriggerAvatar')).not.toBeInTheDocument();
  });

  it('renders avatars initials', () => {
    const max = 3;
    const avatarsInitials = ['JD', 'SP', 'NW'];
    const extraAvatar = list.length - max;

    const { getAllByTestId, getByTestId } = render(<AvatarGroup list={list} max={max} />);
    const avatars = getAllByTestId('DesignSystem-AvatarGroup--Avatar');

    avatars.forEach((avatar, index) => {
      expect(avatar.textContent).toMatch(avatarsInitials[index]);
    });

    expect(getByTestId('DesignSystem-AvatarGroup--TriggerAvatar').textContent).toMatch(`+${extraAvatar}`);
  });

  it('renders popover having the list of full names', () => {
    const max = 3;
    const avatarsInitials = ['Monica Geller', 'Arya Stark', 'Rachel Green'];

    const { getByTestId, getAllByTestId } = render(
      <AvatarGroup list={list} max={max} popoverOptions={{ on: 'click' }} />
    );

    const extraAvatar = getByTestId('DesignSystem-AvatarGroup--TriggerAvatar');
    fireEvent.click(extraAvatar);
    const avatars = getAllByTestId('DesignSystem-AvatarGroup--Text');

    avatars.forEach((avatar, index) => {
      expect(avatar.textContent).toMatch(avatarsInitials[index]);
    });
  });
});

describe('AvatarGroup Component with overwrite class', () => {
  it('overwrite Avatar class', () => {
    const { getByTestId } = render(<AvatarGroup className="AvatarGroupClass" list={list} />);
    expect(getByTestId('DesignSystem-AvatarGroup')).toHaveClass('AvatarGroupClass');
  });
});

describe('AvatarGroup Component with prop: max', () => {
  it('renders avatars with prop: max', () => {
    const max = 3;

    const { getAllByTestId } = render(<AvatarGroup list={list} max={max} />);
    expect(getAllByTestId('DesignSystem-AvatarGroup--Avatar')).toHaveLength(max);
    expect(getAllByTestId('DesignSystem-AvatarGroup--TriggerAvatar')).toHaveLength(1);
  });

  it('renders avatars with default max', () => {
    const defaultMax = 2;

    const { getAllByTestId } = render(<AvatarGroup list={list} />);
    expect(getAllByTestId('DesignSystem-AvatarGroup--Avatar')).toHaveLength(defaultMax);
  });
});

describe('AvatarGroup Component with prop: borderColor', () => {
  it('renders avatars with prop: borderColor', () => {
    const borderColor = 'var(--warning)';
    const border = `2px solid ${borderColor}`;

    const { getByTestId } = render(<AvatarGroup list={list} borderColor={borderColor} />);
    expect(getByTestId('DesignSystem-AvatarGroup--TriggerAvatar')).toHaveStyle(`border: ${border};`);
  });

  it('renders avatars with default borderColor', () => {
    const defaultBorderColor = 2;
    const border = `2px solid ${defaultBorderColor}`;

    const { getByTestId } = render(<AvatarGroup list={list} />);
    expect(getByTestId('DesignSystem-AvatarGroup--TriggerAvatar')).toHaveStyle(`border: ${border};`);
  });
});

describe('AvatarGroup Component with prop: popoverOptions', () => {
  it('renders avatars with prop: popperRenderer', () => {
    const max = 3;
    const popperList = list.slice(max, list.length);
    const popperRenderer = jest.fn();

    const { getByTestId } = render(
      <AvatarGroup list={list} max={max} popoverOptions={{ popperRenderer, on: 'click', dark: false }} />
    );

    const extraAvatar = getByTestId('DesignSystem-AvatarGroup--TriggerAvatar');
    fireEvent.click(extraAvatar);
    expect(popperRenderer).toHaveBeenCalledTimes(1);
    expect(popperRenderer).toHaveBeenCalledWith(popperList);
  });
});
