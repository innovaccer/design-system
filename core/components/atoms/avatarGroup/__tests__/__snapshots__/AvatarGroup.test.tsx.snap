// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup component 
  list: [{"firstName":"John","lastName":"Doe"},{"firstName":"Steven","lastName":"Packton"},{"firstName":"Nancy","lastName":"Wheeler"},{"firstName":"Monica","lastName":"Geller"},{"firstName":"Arya","lastName":"Stark"},{"firstName":"Rachel","lastName":"Green"}]
 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="AvatarGroup d-inline-flex"
        data-test="DesignSystem-AvatarGroup"
      >
        <div
          class="AvatarGroup-item"
          data-test="DesignSystem-AvatarGroup--Avatar"
          style="border-radius: 50%; box-shadow: 0 0 0 var(--spacing-xs) var(--white);"
        >
          <style />
          <div
            class="OutsideClick PopperWrapper-trigger flex-grow-0"
          >
            <span
              class="Avatar Avatar--regular Avatar--accent1"
              data-test="DesignSystem-Avatar"
            >
              <span
                class="Text Text--white Text--medium Text--regular Avatar-content--regular Avatar-content--accent1"
                data-test="DesignSystem-Text"
              >
                JD
              </span>
            </span>
          </div>
        </div>
        <div
          class="AvatarGroup-item"
          data-test="DesignSystem-AvatarGroup--Avatar"
          style="border-radius: 50%; box-shadow: 0 0 0 var(--spacing-xs) var(--white);"
        >
          <style />
          <div
            class="OutsideClick PopperWrapper-trigger flex-grow-0"
          >
            <span
              class="Avatar Avatar--regular Avatar--alert"
              data-test="DesignSystem-Avatar"
            >
              <span
                class="Text Text--white Text--medium Text--regular Avatar-content--regular Avatar-content--alert"
                data-test="DesignSystem-Text"
              >
                SP
              </span>
            </span>
          </div>
        </div>
        <style />
        <div
          class="OutsideClick PopperWrapper-trigger"
        >
          <div
            data-test="DesignSystem-AvatarGroup--TriggerAvatar"
            style="border-radius: 50%; box-shadow: 0 0 0 var(--spacing-xs) var(--white);"
          >
            <span
              class="Avatar Avatar--regular Avatar--secondary Avatar--disabled"
            >
              <span
                class="Text Text--white Text--regular Avatar-content--secondary"
                data-test="DesignSystem-Text"
              >
                +4
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="AvatarGroup d-inline-flex"
      data-test="DesignSystem-AvatarGroup"
    >
      <div
        class="AvatarGroup-item"
        data-test="DesignSystem-AvatarGroup--Avatar"
        style="border-radius: 50%; box-shadow: 0 0 0 var(--spacing-xs) var(--white);"
      >
        <style />
        <div
          class="OutsideClick PopperWrapper-trigger flex-grow-0"
        >
          <span
            class="Avatar Avatar--regular Avatar--accent1"
            data-test="DesignSystem-Avatar"
          >
            <span
              class="Text Text--white Text--medium Text--regular Avatar-content--regular Avatar-content--accent1"
              data-test="DesignSystem-Text"
            >
              JD
            </span>
          </span>
        </div>
      </div>
      <div
        class="AvatarGroup-item"
        data-test="DesignSystem-AvatarGroup--Avatar"
        style="border-radius: 50%; box-shadow: 0 0 0 var(--spacing-xs) var(--white);"
      >
        <style />
        <div
          class="OutsideClick PopperWrapper-trigger flex-grow-0"
        >
          <span
            class="Avatar Avatar--regular Avatar--alert"
            data-test="DesignSystem-Avatar"
          >
            <span
              class="Text Text--white Text--medium Text--regular Avatar-content--regular Avatar-content--alert"
              data-test="DesignSystem-Text"
            >
              SP
            </span>
          </span>
        </div>
      </div>
      <style />
      <div
        class="OutsideClick PopperWrapper-trigger"
      >
        <div
          data-test="DesignSystem-AvatarGroup--TriggerAvatar"
          style="border-radius: 50%; box-shadow: 0 0 0 var(--spacing-xs) var(--white);"
        >
          <span
            class="Avatar Avatar--regular Avatar--secondary Avatar--disabled"
          >
            <span
              class="Text Text--white Text--regular Avatar-content--secondary"
              data-test="DesignSystem-Text"
            >
              +4
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
